drivers/wifi/wifi_winc1500.c

```diff
@@ -0,0 +1,924 @@
+	if (ret) {
+		SYS_LOG_ERR("bind error %d %s!",
+			    ret, socket_message_to_string(ret));
+		return ret;
+	}
+
+	if (k_sem_take(&w1500_data.socket_data[socket].wait_sem,
+		       WINC1500_BIND_TIMEOUT)) {
+		SYS_LOG_ERR("bind error timeout expired");
+		return -ETIMEDOUT;
+	}
+
+	return w1500_data.socket_data[socket].ret_code;
+}
+
+/**
+ * This function is called when user wants to mark the socket
+ * to be a listening one.
+ */
+static int winc1500_listen(struct net_context *context, int backlog)
+{
+	SOCKET socket = (int)context->user_data;
+	int ret;
+
+	ret = listen((int)context->user_data, backlog);
+	if (ret) {
+		SYS_LOG_ERR("listen error %d %s!",
+			    ret, socket_error_string(ret));
+		return ret;
+	}
+
+	if (k_sem_take(&w1500_data.socket_data[socket].wait_sem,
+		       WINC1500_LISTEN_TIMEOUT)) {
+		return -ETIMEDOUT;
+	}
+
+	return w1500_data.socket_data[socket].ret_code;
+}
+
+/**
+ * This function is called when user wants to create a connection
+ * to a peer host.
+ */
+static int winc1500_connect(struct net_context *context,
+			    const struct sockaddr *addr,
+			    socklen_t addrlen,
+			    net_context_connect_cb_t cb,
+			    int32_t timeout,
+			    void *user_data)
+{
+	SOCKET socket = (int)context->user_data;
+	int ret;
+
+	w1500_data.socket_data[socket].connect_cb = cb;
+	w1500_data.socket_data[socket].connect_user_data = user_data;
+	w1500_data.socket_data[socket].ret_code = 0;
+
+	ret = connect(socket, (struct sockaddr *)addr, addrlen);
+	if (ret) {
+		SYS_LOG_ERR("connect error %d %s!",
+			    ret, socket_error_string(ret));
+		return ret;
+	}
+
+	if (timeout != 0 &&
+	    k_sem_take(&w1500_data.socket_data[socket].wait_sem, timeout)) {
+		return -ETIMEDOUT;
+	}
+
+	return w1500_data.socket_data[socket].ret_code;
+}
+
+/**
+ * This function is called when user wants to accept a connection
+ * being established.
+ */
+static int winc1500_accept(struct net_context *context,
+			   net_tcp_accept_cb_t cb,
+			   int32_t timeout,
+			   void *user_data)
+{
+	SOCKET socket = (int)context->user_data;
+	int ret;
+
+	w1500_data.socket_data[socket].accept_cb = cb;
+
+	ret = accept(socket, NULL, 0);
+	if (ret) {
+		SYS_LOG_ERR("accept error %d %s!",
+			    ret, socket_error_string(ret));
+		return ret;
+	}
+
+	if (timeout) {
+		if (k_sem_take(&w1500_data.socket_data[socket].wait_sem,
+			       timeout)) {
+			return -ETIMEDOUT;
+		}
+	} else {
+		k_sem_take(&w1500_data.socket_data[socket].wait_sem,
+			   K_FOREVER);
+	}
+
+	return w1500_data.socket_data[socket].ret_code;
+}
+
+/**
+ * This function is called when user wants to send data to peer host.
+ */
+static int winc1500_send(struct net_pkt *pkt,
+			 net_context_send_cb_t cb,
+			 int32_t timeout,
+			 void *token,
+			 void *user_data)
+{
+	struct net_context *context = pkt->context;
+	SOCKET socket = (int)context->user_data;
+	bool first_frag;
+	struct net_buf *frag;
+	int ret;
+
+	w1500_data.socket_data[socket].send_cb = cb;
```

Comment:
Trying to follow the logic... Where does the send_cb get called to report the actual number of bytes sent?  Is it in a different file?

Reply:
original PR did not seem to implement handling SOCKET_MSG_SEND in winc1500_socket_cb()

Quite many stuff were required to be fixed to get that PR in an okeyish state.
further fix will come in future patches, there is a lot to test and fix around that driver and NET_OFFLOAD